lib/externalGraderResults.js

```diff
@@ -11,129 +10,116 @@ const externalGradingSocket = require('./externalGradingSocket');
+          try {
+            const parsedMessage = JSON.parse(message.Body);
+            const receiptHandle = message.ReceiptHandle;
+
+            await processMessage(parsedMessage);
+
+            await sqs
+              .deleteMessage({
+                QueueUrl: queueUrl,
+                ReceiptHandle: receiptHandle,
+              })
+              .promise();
+          } catch (err) {
+            logger.error('Error processing external grader results', err);
+            Sentry.captureException(err);
+          }
+        })
       );
-    });
-  });
+    }
+  })();
 };
 
-function loadQueueUrl(sqs, callback) {
-  if (QUEUE_URL !== null) {
-    callback(null);
-  } else {
-    logger.verbose(
-      `External grading results queue ${config.externalGradingResultsQueueName}: getting URL...`
-    );
-    const params = {
-      QueueName: config.externalGradingResultsQueueName,
-    };
-    sqs.getQueueUrl(params, (err, data) => {
-      if (ERR(err, callback)) return;
-      QUEUE_URL = data.QueueUrl;
-      logger.verbose(
-        `External grading results queue ${config.externalGradingResultsQueueName}: got URL ${QUEUE_URL}`
-      );
-      callback(null);
-    });
+module.exports.stop = async function () {
+  if (abortController.signal.aborted) {
+    throw new Error('Already stopped');
   }
-}
+  abortController.abort();
+
+  // The main work loop will resolve this deferred promise when it's finished
+  // with any current processing.
+  await processingFinished.promise();
+};
 
-function logAtTimeout(jobId, logBundle) {
-  logger.info('externalGradingResults timeout', { jobId, logBundle });
+/**
+ *
+ * @param {import('aws-sdk').SQS} sqs
+ * @returns {Promise<string>} The URL of the results queue.
+ */
+async function loadQueueUrl(sqs) {
+  logger.verbose(
+    `External grading results queue ${config.externalGradingResultsQueueName}: getting URL...`
+  );
+  const data = await sqs
+    .getQueueUrl({ QueueName: config.externalGradingResultsQueueName })
+    .promise();
+  const queueUrl = data.QueueUrl;
+  logger.verbose(
+    `External grading results queue ${config.externalGradingResultsQueueName}: got URL ${queueUrl}`
+  );
+  return queueUrl;
 }
 
-function processMessage(data, callback) {
-  let jobId;
-  try {
-    jobId = Number.parseInt(data.jobId);
-    if (Number.isNaN(jobId)) {
-      throw new Error();
-    }
-  } catch (e) {
-    return callback(error.makeWithData('Message does not contain a valid grading job id.', data));
+async function processMessage(data) {
+  const jobId = Number.parseInt(data.jobId);
+  if (Number.isNaN(jobId)) {
+    throw error.makeWithData('Message does not contain a valid grading job id.', data);
   }
 
-  var logBundle = [];
-  var loggingTimeout = setTimeout(logAtTimeout, 10000, jobId, logBundle);
+  const logBundle = [];
+  const loggingTimeout = setTimeout(() => {
+    logger.info('externalGradingResults timeout', { jobId, logBundle });
+  }, 10000);
```

Comment:
What exactly is this log bundle stuff? I can see it was introduced in #1868, but that doesn't include any description of the problem we're trying to solve. Were we experiencing results that took a long time to process? Do you think this is still necessary?

Reply:
Done in 302bacd.